internal/controller/collection/reconciler.go

```diff
@@ -0,0 +1,277 @@
+		if err := r.client.Status().Update(ctx, collection); err != nil {
+			return ctrl.Result{}, errors.Wrap(err, errStatusUpdate)
+		}
+		return ctrl.Result{}, errors.Wrap(err, "error listing objects")
+	}
+
+	// Create observed-only Objects for all found items
+	refs := sets.New[v1alpha1.ObservedObjectReference]()
+	for _, o := range objects.Items {
+		log.Debug("creating observed object for the matched item", "gvk", o.GroupVersionKind(), "name", o.GetName())
+		name, err := r.observedObjectName(collection, &o) //#nosec G601
+		if err != nil {
+			collection.Status.SetConditions(xpv1.ReconcileError(err))
+			if err := r.client.Status().Update(ctx, collection); err != nil {
+				return ctrl.Result{}, errors.Wrap(err, errStatusUpdate)
+			}
+			return ctrl.Result{}, errors.Wrap(err, fmt.Sprintf("error generating name for observed object, matched object: %v", o))
+		}
+
+		// Create patch
+		po, err := observedObjectPatch(name, o, collection)
+		if err != nil {
+			collection.Status.SetConditions(xpv1.ReconcileError(err))
+			if err := r.client.Status().Update(ctx, collection); err != nil {
+				return ctrl.Result{}, errors.Wrap(err, errStatusUpdate)
+			}
+			return ctrl.Result{}, errors.Wrap(err, fmt.Sprintf("error generating patch for matched object %v", o))
+		}
+
+		if err := r.client.Patch(ctx, po, client.Apply, fieldOwner, client.ForceOwnership); err != nil {
+			collection.Status.SetConditions(xpv1.ReconcileError(err))
+			if err := r.client.Status().Update(ctx, collection); err != nil {
+				return ctrl.Result{}, errors.Wrap(err, errStatusUpdate)
+			}
+			return ctrl.Result{}, errors.Wrap(err, "cannot create observed object")
+		}
+
+		log.Debug("created observed object", "name", po.GetName())
+		refs.Insert(v1alpha1.ObservedObjectReference{Name: name})
+	}
+
+	// Remove objects that either do not exist anymore or are no match
+	oldRefs := sets.New[v1alpha1.ObservedObjectReference](collection.Status.Objects...)
+	for _, or := range oldRefs.Difference(refs).UnsortedList() {
+		o := &v1alpha2.Object{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: or.Name,
+			}}
+		if err := r.client.Delete(ctx, o); err != nil {
+			collection.Status.SetConditions(xpv1.ReconcileError(err))
+			if err := r.client.Status().Update(ctx, collection); err != nil {
+				return ctrl.Result{}, errors.Wrap(err, errStatusUpdate)
+			}
+			return ctrl.Result{}, errors.Wrap(err, fmt.Sprintf("cannot delete observed object %s", or.Name))
+		}
+	}
+	collection.Status.SetConditions(xpv1.ReconcileSuccess(), xpv1.Available())
+
+	// Update references only if more or less objects got matched since the last reconciliation.
+	if !refs.Equal(oldRefs) {
+		collection.Status.Objects = refs.UnsortedList()
+	}
+
+	return ctrl.Result{RequeueAfter: r.pollInterval()}, r.client.Status().Update(ctx, collection)
+}
+
+func observedObjectName(collection client.Object, matchedObject client.Object) (string, error) {
+	h := md5.New() //#nosec G401 -- used only for unique name generation
```

Comment:
I like the idea of leveraging MATCHED_OBJECT_UUID and avoiding extra calls to API server 👍

Reply:
In order to make it backup-friendly, I have update the convention to:

```
${COLLECTION_NAME}-${MATCHED_OBJECT_KIND}-${MATCHED_OBJECT_NAMESPACE}-${MATCHED_OBJECT_NAME}
```

if the generate name is longer than 253 chars, we generate md5 hash of it and use it instead.

Reply:
We have a couple of existing ways to name automatically generated child resources:

* For XRs and composed resources, we use the NameGenerator, that mimics `metadata.generateName` (see [here](https://github.com/crossplane/crossplane/blob/v1.15.1/internal/names/generate.go#L68))
* For composition revisions, we use the composition name suffixed with a truncated SHA256 hash of the composition (see [here](https://github.com/crossplane/crossplane/blob/v1.15.1/internal/controller/apiextensions/composition/revision.go#L30))
* For package revisions, we use the package name suffixed with a the OCI digest (i.e. truncated SHA256 hash) of the package (see [here](https://github.com/crossplane/crossplane/blob/v1.15.1/internal/controller/pkg/manager/revisioner.go#L87))

I'd prefer not to avoid another naming scheme, unless there's strong reason to. I'd especially prefer to avoid introducing a naming scheme with two variants (e.g. sometimes one format, sometimes md5 hashes).

Can we just use `metadata.generateName`, to let the API server name these objects? This would be aesthetically identical to the composition NameGenerator, and avoid any concerns about extra calls to the API server. (Although for the record, I don't share those concerns.)

I don't think we need the NameGenerator for cases like this where an `ObservedObjectCollection` will always create one kind of resource - `Object`. It would be relatively cheap for this controller to:

* Just use `metadata.generateName`
* Proactively check for leaked resources by listing `Object` filtered by `kubernetes.crossplane.io/owned-by-collection` before creating anything new

I actually think this applies to the claim controller too - raised https://github.com/crossplane/crossplane/issues/5535 to track that.

Reply:
> Can we just use metadata.generateName, to let the API server name these objects? 

We cannot unfortunately, because a server-side apply patch must be named, i.e. `metadata.name` should be set. SSA is used because we want to propagate correctly (and remove too) annotations/labels that users can add through `spec.objectTemplate` to generated Objects.

* The proposed scheme is very simple and produces deterministic and always available names that remain stable across arbitrary number of backup/restore cycles both on crossplane and remote clusters.
* It falls back to md5 hash only if the name is longer than 253 chars (and we agree that it gonna happen rarely)

What is the advantage of using random names when we have the deterministic solution?
* There is always a chance for collision
* In each reconcile loop, we need to (1) create new objects if there are more matches, (2) remove objects not matched anymore, and (3) update all existing objects because `spec.objectTemplate` labels/annotations could change. Having the deterministic names helps here a lot to keep the logic simple.

> Proactively check for leaked resources by listing Object filtered by kubernetes.crossplane.io/owned-by-collection before creating anything new

In order to detect leakage, we need to come up with an identifier that can be found/computed on the fresh query result. If we use random Object names, we need to discover which item in the results correspond to which Object. Eventually we are going to build an index over Objects, but we could use that index key as Object name and save us one pass over them.

BTW, if we apply the random name generation starting from a claim `foo`, this is what we are going to get on collection objects:

* claim name: `foo`
* composite name: `foo-1dgft`
* observed object collection: `foo-1dgft-tr451`
* observed object collection member1: `foo-1dgft-tr451-ptrqw`
* observed object collection member1: `foo-1dgft-tr451-qwrty`

In the proposed naming schema, the last two names could look like `foo-1dgft-tr451-configmap-example1-cfg1` and `foo-1dgft-tr451-configmap-example2-cfg1`.

Reply:
Not sure I understand all the context. What I see are two approaches:
1. two phase commit with the simple namer (including conflict recovery). This means that we store the name mapping in the collection before creating an object.
2. deterministic names. This does not need to store names. We assume the deterministic names are unique enough such that in our life-time there is never a conflict.

So it depends on whether we want to store names. I tend to prefer the deterministic approach as it is easier and has less write to the apiserver.

Reply:
> two phase commit with the simple namer (including conflict recovery). This means that we store the name mapping in the collection before creating an object.

We use SSA (Patch) so we are not able to detect conflict - if object already exists, we are going to update it. Crossplane has the same issue: https://github.com/crossplane/crossplane/blob/master/internal/controller/apiextensions/composite/composition_functions.go#L473

Reply:
The spirit of my feedback: The answer isn't always "what is the best thing we could do, in isolation, for this new feature".

This isn't a greenfield context. Crossplane already has two patterns to name child resources - `NameGenerator` and truncated SHA256 hashes. If we can use one of those, we should strongly consider it. Consistency matters. Consistent UX and aesthetics for end users, and consistent implementations for everyone (users and maintainers) to reason about.

If every time we face the same problem, we solve it a little bit differently (even if it's better!) we end up with a system that's more surprising and harder to reason about overall.

> Having the deterministic names helps here a lot to keep the logic simple.

That makes sense. One of the two existing patterns I called out above is deterministic - truncated SHA256 hashes.

Could we take the deterministic data you're already using (matched object GVK + namespace + name), SHA256 hash it, and truncate it to 7 characters? So for an `ObservedObjectCollection` named `jobs`, each child `Object` would have a name like `jobs-d35ecef`.

Pros:

* Deterministic names
* Uses an existing, established pattern
* Short, and collision resistant enough for at least hundreds of children of one `ObservedObjectCollection`
* Consistent format (doesn't toggle between human readable and a hash depending on input data length)

Cons:

* Not human readable

Not being human readable is a little unfortunate, but 'm not overly worried about it. The GVK, namespace and name of the wrapped object are all print columns of the `Object` CRD and will thus show up in `kubectl get objects -l kubernetes.crossplane.io/owned-by-collection=jobs`.

Reply:
> Could we take the deterministic data you're already using (matched object GVK + namespace + name), SHA256 hash it, and truncate it to 7 characters?

Sure, pushed the change. By looking at your earlier comments, my understanding was that you favor `NameGenerator` option.

BTW, from math point of view, using SHA256 over MD5 to compute hashes does not reduce chances for name collision, because we take just first 56 bits.